[SPARK-19556][core] Do not encrypt block manager data in memory.

vanzin opened the pull request:
This change modifies the way block data is encrypted to make the more
common cases faster, while penalizing an edge case. As a side effect
of the change, all data that goes through the block manager is now
encrypted only when needed, including the previous path (broadcast
variables) where that did not happen.

The way the change works is by not encrypting data that is stored in
memory; so if a serialized block is in memory, it will only be encrypted
once it is evicted to disk.

The penalty comes when transferring that encrypted data from disk. If the
data ends up in memory again, it is as efficient as before; but if the
evicted block needs to be transferred directly to a remote executor, then
there's now a performance penalty, since the code now uses a custom
FileRegion implementation to decrypt the data before transferring.

This also means that block data transferred between executors now is
not encrypted (and thus relies on the network library encryption support
for secrecy). Shuffle blocks are still transferred in encrypted form,
since they're handled in a slightly different way by the code. This also
keeps compatibility with existing external shuffle services, which transfer
encrypted shuffle blocks, and avoids having to make the external service
aware of encryption at all.

The serialization and deserialization APIs in the SerializerManager now
do not do encryption automatically; callers need to explicitly wrap their
streams with an appropriate crypto stream before using those.

As a result of these changes, some of the workarounds added in SPARK-19520
are removed here.

Testing: a new trait ("EncryptionFunSuite") was added that provides an easy
way to run a test twice, with encryption on and off; broadcast, block manager
and caching tests were modified to use this new trait so that the existing
tests exercise both encrypted and non-encrypted paths. I also ran some
applications with encryption turned on to verify that they still work,
including streaming tests that failed without the fix for SPARK-19520.

Commit:
[SPARK-19556][core] Do not encrypt block manager data in memory.

This change modifies the way block data is encrypted to make the more
common cases faster, while penalizing an edge case. As a side effect
of the change, all data that goes through the block manager is now
encrypted only when needed, including the previous path (broadcast
variables) where that did not happen.

The way the change works is by not encrypting data that is stored in
memory; so if a serialized block is in memory, it will only be encrypted
once it is evicted to disk.

The penalty comes when transferring that encrypted data from disk. If the
data ends up in memory again, it is as efficient as before; but if the
evicted block needs to be transferred directly to a remote executor, then
there's now a performance penalty, since the code now uses a custom
FileRegion implementation to decrypt the data before transferring.

This also means that block data transferred between executors now is
not encrypted (and thus relies on the network library encryption support
for secrecy). Shuffle blocks are still transferred in encrypted form,
since they're handled in a slightly different way by the code. This also
keeps compatibility with existing external shuffle services, which transfer
encrypted shuffle blocks, and avoids having to make the external service
aware of encryption at all.

Another change in the disk store is that it now stores a tiny metadata
file next to the file holding the block data; this is needed to accurately
account for the decrypted block size, which may be significantly different
from the size of the encrypted file on disk.

The serialization and deserialization APIs in the SerializerManager now
do not do encryption automatically; callers need to explicitly wrap their
streams with an appropriate crypto stream before using those.

As a result of these changes, some of the workarounds added in SPARK-19520
are removed here.

Testing: a new trait ("EncryptionFunSuite") was added that provides an easy
way to run a test twice, with encryption on and off; broadcast, block manager
and caching tests were modified to use this new trait so that the existing
tests exercise both encrypted and non-encrypted paths. I also ran some
applications with encryption turned on to verify that they still work,
including streaming tests that failed without the fix for SPARK-19520.

SparkQA commented:
**[Test build #74555 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/74555/testReport)** for PR 17295 at commit [`3aa752f`](https://github.com/apache/spark/commit/3aa752f9becdfe0e35a47d731736d942e3e5b3bf).

mridulm commented:
I have not looked at the implementation in detail, but can you comment on why the change w.r.t plain text block data to remote executor ? Isn't it not simpler to transmit block contents in encrypted format without decryption ?

Remote fetch of RDD blocks is not uncommon (for any task other than PROCESS_LOCAL); and I wanted to better understand why this is required.

vanzin commented:
> Isn't it not simpler to transmit block contents in encrypted format without decryption?

First, keep in mind that there's no metadata that tells the receiver whether a block is encrypted or not. This means that methods like `BlockManager.get`, which can read block data from either local or remote sources, need to return data that is either always encrypted or always not encrypted for the same block ID.

This leaves two choices:

- encrypt the data in all stores (memory & disk); this is what the current code does, and it requires all code that uses the BlockManager to have to deal with encryption. This is what caused SPARK-19520, and I filed SPARK-19556 to cover yet another case of a code path that did not do the right thing when encryption is enabled.

- make all non-shuffle block data read from the BlockManager not encrypted. This means non-shuffle code calling the BlockManager does not have to care about encryption, since it will always read unencrypted data, and can always put unencrypted data in the BlockManager and it will be encrypted when needed (a.k.a. when writing to disk).

> Remote fetch of RDD blocks is not uncommon

That's fine. This change makes the data read from the BlockManager instance not encrypted. But when transmitting the data over to another executor, there's RPC-level encryption (`spark.authenticate.enableSaslEncryption` or `spark.network.crypto.enabled`), which means the data is still encrypted on the wire.

mridulm commented:
> First, keep in mind that there's no metadata that tells the receiver whether a block is encrypted or not. This means that methods like BlockManager.get, which can read block data from either local or remote sources, need to return data that is either always encrypted or always not encrypted for the same block ID.

This can be solved by tagging the block data with a prefix byte - we do something similar for MapStatus (direct or broadcast).

SparkQA commented:
**[Test build #74555 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/74555/testReport)** for PR 17295 at commit [`3aa752f`](https://github.com/apache/spark/commit/3aa752f9becdfe0e35a47d731736d942e3e5b3bf).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/74555/
Test PASSed.

mridulm commented:
Just to be clear, I would prefer if we consistently did things - either encrypt all blocks while transferring (irrespective of sasl being enabled or not); or depend only on sasl for channel encryption.
But given this was what it currently is, I am not sure if it was by design or accident; and what the tradeoff's for ensuring consistency is.

(The workaround is, what I mentioned above, tagging)

vanzin commented:
> This can be solved by tagging the block data with a prefix byte

Sure, it could be solved in different ways. I just happened to prefer the one in this patch, since I think it's less intrusive; if you look closely, the majority of changes are in a single class (`DiskStore`), and there's mostly minor adjustments in other places.

vanzin commented:
> Just to be clear, I would prefer if we consistently did things - either encrypt all blocks while transferring (irrespective of sasl being enabled or not); or depend only on sasl for channel encryption.

Not really sure what you mean here. But transferring encrypted data without RPC encryption is not really secure, since the encryption key is transferred to executors using an RPC. There's even a warning message if RPC encryption is not on and you enable disk encryption.

Shuffle is a different beast - I explain why the shuffle blocks are transferred in encrypted form in the PR description.

mridulm commented:
> Not really sure what you mean here. But transferring encrypted data without RPC encryption is not really secure, since the encryption key is transferred to executors using an RPC. There's even a warning message if RPC encryption is not on and you enable disk encryption.

Good point, I overlooked that.
So to summarize, after this change, RDD block's transferred will always be in plain text; with an implicit requirement that rpc encryption is strongly preferred to be enabled.
Is there any case where it is transfered in encrypted form in supported cases ? (cases being: broadcast, rdd block transfer, replication, anything else ?)
I wanted to ensure I understand what the final expected behavior/state would be, and how consistent we will become.

I agree about shuffle being special case'd; I was looking at only non-shuffle blocks.

vanzin commented:
> Is there any case where it is transfered in encrypted form in supported cases ?

No, with these changes, only shuffle data is transferred in encrypted form.

sameeragarwal commented:
cc @cloud-fan @ueshin

cloud-fan commented:
> The penalty comes when transferring that encrypted data from disk. If the
data ends up in memory again, it is as efficient as before; but if the
evicted block needs to be transferred directly to a remote executor, then
there's now a performance penalty, since the code now uses a custom
FileRegion implementation to decrypt the data before transferring.

What's the actual difference? previously we transfer encrypted data?

vanzin commented:
> What's the actual difference? previously we transfer encrypted data?

Yes. The previous version of the code would transfer the encrypted file over to the receiver, and the encrypted data for serialized blocks would also be stored in `MemoryStore` (and then decrypted on every use). That means the files could just be mmap'ed for transfer, which is faster than the `ReadableByteChannel` path even without encryption in the picture. (If you consider the previous code had to decrypt from the `MemoryStore` on every read, you can end up with better performance overall with this patch.)

But this caused all the other issues with making the `BlockManager` harder to use when encryption was on, so I think this is a better solution.

cloud-fan commented:
makes sense. one more question, ideally, shall we also transfer shuffle blocks after decryption?

cloud-fan commented on core/src/main/scala/org/apache/spark/security/CryptoStreamUtils.scala:
```diff
@@ -63,12 +83,40 @@ private[spark] object CryptoStreamUtils extends Logging {
-    new CryptoInputStream(transformationStr, properties, is,
-      new SecretKeySpec(key, "AES"), new IvParameterSpec(iv))
+    var read = 0
+    while (read < iv.length) {
```
what does this while loop do?

vanzin replied:
It avoids issues with short reads. It's unlikely to happen but I always write read code like this to be safe.

zsxwing replied:
Yeah, you can just use `ByteStreams.readFully(is, iv)`.

vanzin replied:
Ah, missed that one. +1 for shorter code.

cloud-fan commented on core/src/main/scala/org/apache/spark/security/CryptoStreamUtils.scala:
```diff
@@ -102,4 +150,34 @@ private[spark] object CryptoStreamUtils extends Logging {
+  /**
+   * This class is a workaround for CRYPTO-125, that forces all bytes to be written to the
+   * underlying channel. Since the callers of this API are using blocking I/O, there are no
+   * concerns with regards to CPU usage here.
```
is it a separated bug fix?

vanzin replied:
No. As the comment states, it's a workaround for a bug in the commons-crypto library, which would affect the code being added.

cloud-fan commented on core/src/main/scala/org/apache/spark/storage/BlockManager.scala:
```diff
@@ -1235,7 +1251,7 @@ private[spark] class BlockManager(
           peer.executorId,
           blockId,
-          new NettyManagedBuffer(data.toNetty),
+          new BlockManagerManagedBuffer(blockInfoManager, blockId, data.toManagedBuffer()),
```
why this change?

cloud-fan commented on core/src/main/scala/org/apache/spark/storage/DiskStore.scala:
```diff
@@ -73,55 +86,219 @@ private[spark] class DiskStore(conf: SparkConf, diskManager: DiskBlockManager) e
+      assert(blockSize <= Int.MaxValue, "Block is too large to be wrapped in a byte buffer.")
+      val is = toInputStream()
+      try {
+        ByteBuffer.wrap(ByteStreams.toByteArray(is))
```
will we read all data out here?

vanzin replied:
There's a comment explaining it a few lines above...

vanzin commented:
>  shall we also transfer shuffle blocks after decryption?

No. That's explained in the PR description.

mridulm commented on core/src/main/scala/org/apache/spark/storage/DiskStore.scala:
```diff
@@ -73,55 +86,219 @@ private[spark] class DiskStore(conf: SparkConf, diskManager: DiskBlockManager) e
+        remaining -= chunkSize
+
+        while (chunk.remaining() > 0) {
+          source.read(chunk)
```
as mentioned above, needs EOF error handling.

mridulm commented on core/src/main/scala/org/apache/spark/storage/DiskStore.scala:
```diff
@@ -73,55 +86,219 @@ private[spark] class DiskStore(conf: SparkConf, diskManager: DiskBlockManager) e
+    written
+  }
+
+  override def deallocate(): Unit = source.close()
```
release buffer as well.

vanzin replied:
`StorageUtils.dispose` specifically checks for mapped buffers, which is not the case here. It could be changed, but in this case I wonder if it's necessary or if waiting for GC is good enough.

mridulm reviewed (commented):
I did an initial pass and added some comments/queries. Overall, as I mentioned earlier, I like the fact that we have a more consistent approach to transfering data.
Thanks for the work @vanzin !

vanzin commented on core/src/main/scala/org/apache/spark/security/CryptoStreamUtils.scala:
```diff
@@ -102,4 +150,34 @@ private[spark] object CryptoStreamUtils extends Logging {
   }
+
+  /**
+   * This class is a workaround for CRYPTO-125, that forces all bytes to be written to the
```
There's a pretty nasty workaround for it in the network library... (the non-blocking workaround is a lot worse than this.)

vanzin commented on core/src/main/scala/org/apache/spark/storage/DiskBlockManager.scala:
```diff
@@ -34,6 +34,8 @@ import org.apache.spark.util.{ShutdownHookManager, Utils}
  */
 private[spark] class DiskBlockManager(conf: SparkConf, deleteFilesOnStop: Boolean) extends Logging {
 
+  private val METADATA_FILE_SUFFIX = ".meta"
```
Hmm, good point... there's currently no metadata kept in the `DiskStore` class, but then this shouldn't be a lot of data.

vanzin commented on core/src/main/scala/org/apache/spark/storage/DiskStore.scala:
```diff
@@ -17,48 +17,61 @@
     } finally {
       try {
-        Closeables.close(fileOutputStream, threwException)
+        Closeables.close(out, threwException)
```
This was the previous behavior, but well, doesn't hurt to fix it.

vanzin commented on core/src/main/scala/org/apache/spark/storage/DiskStore.scala:
```diff
@@ -73,55 +86,219 @@ private[spark] class DiskStore(conf: SparkConf, diskManager: DiskBlockManager) e
-        channel.close()
-      }
+    put(blockId) { channel =>
+      bytes.writeFully(channel)
```
The channel is owned by the code in the `put` method, which does that.

vanzin commented on core/src/main/scala/org/apache/spark/storage/DiskStore.scala:
```diff
@@ -73,55 +86,219 @@ private[spark] class DiskStore(conf: SparkConf, diskManager: DiskBlockManager) e
+      }.getOrElse(out)
+    } catch {
+      case e: Exception =>
+        out.close()
```
There might be exceptions specific for the commons-crypto library being thrown.

Commit:
Feedback: style.

Commit:
Add helper method to read buffer from channel.

Commit:
Store DiskStore block sizes in memory.

Commit:
Fix BlockData lifecycle management.

This wasn't really caused by the new code, but by old code that was not
consistent in its disposal of BlockManager buffers. This commit fixes the
places where the underlying buffer was just left for the GC instead of
being explicitly disposed.

vanzin commented on core/src/main/scala/org/apache/spark/storage/BlockManager.scala:
```diff
@@ -56,6 +57,43 @@ private[spark] class BlockResult(
+
+  override def toByteBuffer(allocator: Int => ByteBuffer): ChunkedByteBuffer = {
+    buffer.copy(allocator)
+  }
```
So I had traced through that stuff 2 or 3 times, and now I did it again and I think I finally understood all that's going on. Basically, the old code was really bad at explicitly disposing of the buffers, meaning a bunch of paths (like the ones that used managed buffers) didn't bother to do it and just left the work to the GC.

I changed the code a bit to make the dispose more explicit and added comments in a few key places.

SparkQA commented:
**[Test build #74905 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/74905/testReport)** for PR 17295 at commit [`1428fcd`](https://github.com/apache/spark/commit/1428fcd952ddcdb29a561d8c1c90d4f820955c15).

Commit:
Remove block from size map in DiskStore.remove().

SparkQA commented:
**[Test build #74906 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/74906/testReport)** for PR 17295 at commit [`1b2a3e4`](https://github.com/apache/spark/commit/1b2a3e48e07118b03e9607bf0dc99cf53ae4d678).

SparkQA commented:
**[Test build #74906 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/74906/testReport)** for PR 17295 at commit [`1b2a3e4`](https://github.com/apache/spark/commit/1b2a3e48e07118b03e9607bf0dc99cf53ae4d678).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/74906/
Test FAILed.

vanzin commented:
retest this please

SparkQA commented:
**[Test build #74911 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/74911/testReport)** for PR 17295 at commit [`1b2a3e4`](https://github.com/apache/spark/commit/1b2a3e48e07118b03e9607bf0dc99cf53ae4d678).

SparkQA commented:
**[Test build #74905 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/74905/testReport)** for PR 17295 at commit [`1428fcd`](https://github.com/apache/spark/commit/1428fcd952ddcdb29a561d8c1c90d4f820955c15).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/74905/
Test PASSed.

SparkQA commented:
**[Test build #74911 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/74911/testReport)** for PR 17295 at commit [`1b2a3e4`](https://github.com/apache/spark/commit/1b2a3e48e07118b03e9607bf0dc99cf53ae4d678).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/74911/
Test FAILed.

Commit:
Merge branch 'master' into SPARK-19556

SparkQA commented:
**[Test build #74989 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/74989/testReport)** for PR 17295 at commit [`6848a59`](https://github.com/apache/spark/commit/6848a592df16f778afebf26b19f81f6f23b80aa4).

SparkQA commented:
**[Test build #74989 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/74989/testReport)** for PR 17295 at commit [`6848a59`](https://github.com/apache/spark/commit/6848a592df16f778afebf26b19f81f6f23b80aa4).
 * This patch **fails to build**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/74989/
Test FAILed.

Commit:
Fix compilation after merge.

SparkQA commented:
**[Test build #74991 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/74991/testReport)** for PR 17295 at commit [`107e3e7`](https://github.com/apache/spark/commit/107e3e72e81d2c7813d832d3e9c2beab89e01379).

SparkQA commented:
**[Test build #74991 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/74991/testReport)** for PR 17295 at commit [`107e3e7`](https://github.com/apache/spark/commit/107e3e72e81d2c7813d832d3e9c2beab89e01379).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/74991/
Test FAILed.

Commit:
Need to use ChunkedByteBuffer.unmap() instead of dispose().

SparkQA commented:
**[Test build #74997 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/74997/testReport)** for PR 17295 at commit [`6bda670`](https://github.com/apache/spark/commit/6bda6701bf0c266047a5fa81fd29f4fb826728c7).

SparkQA commented:
**[Test build #74997 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/74997/testReport)** for PR 17295 at commit [`6bda670`](https://github.com/apache/spark/commit/6bda6701bf0c266047a5fa81fd29f4fb826728c7).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/74997/
Test PASSed.

cloud-fan commented on core/src/main/scala/org/apache/spark/broadcast/TorrentBroadcast.scala:
```diff
@@ -219,18 +219,22 @@ private[spark] class TorrentBroadcast[T: ClassTag](obj: T, id: Long)
+            }
+            obj
+          } finally {
+            blocks.foreach(_.dispose())
```
ah good catch! we should dispose the blocks here

cloud-fan commented on core/src/main/scala/org/apache/spark/security/CryptoStreamUtils.scala:
```diff
@@ -48,12 +51,30 @@ private[spark] object CryptoStreamUtils extends Logging {
+    val params = new CryptoParams(key, sparkConf)
+    val iv = createInitializationVector(params.conf)
+    val buf = ByteBuffer.wrap(iv)
+    while (buf.hasRemaining()) {
```
is there any possibility this may be an infinite loop?

cloud-fan replied:
actually this logic is same with `CryptoHelperChannel`. Shall we create `CryptoHelperChannel` first and simply call `helper.write(buf)` here?

vanzin replied:
No, there's no infinite loop here, because a failure would cause an exception. Yeah, using the helper should work too.

cloud-fan commented on core/src/main/scala/org/apache/spark/security/CryptoStreamUtils.scala:
```diff
@@ -63,12 +84,27 @@ private[spark] object CryptoStreamUtils extends Logging {
+      key: Array[Byte]): ReadableByteChannel = {
+    val iv = new Array[Byte](IV_LENGTH_IN_BYTES)
+    val buf = ByteBuffer.wrap(iv)
+    JavaUtils.readFully(channel, buf)
```
why not use `ByteStreams.readFully`? the `buf` is not used else where

vanzin replied:
There's no `ByteStreams.readFully` for `ReadableByteChannel` that I'm aware of.

cloud-fan commented on core/src/main/scala/org/apache/spark/serializer/SerializerManager.scala:
```diff
@@ -167,30 +167,26 @@ private[spark] class SerializerManager(
     val byteStream = new BufferedOutputStream(outputStream)
     val autoPick = !blockId.isInstanceOf[StreamBlockId]
     val ser = getSerializer(implicitly[ClassTag[T]], autoPick).newInstance()
-    ser.serializeStream(wrapStream(blockId, byteStream)).writeAll(values).close()
```
the `wrapStream` and `wrapForEncryption` methods can be removed from this class

vanzin replied:
They're still used in a bunch of places.

cloud-fan commented on core/src/main/scala/org/apache/spark/storage/BlockManager.scala:
```diff
@@ -56,6 +57,44 @@ private[spark] class BlockResult(
+
+  def toChunkedByteBuffer(allocator: Int => ByteBuffer): ChunkedByteBuffer
+
+  def toByteBuffer(): ByteBuffer
```
it will be great to add some document for these 4 methods about when they will be called.

vanzin replied:
I added scaladoc for `toNetty()`, but the others seem self-explanatory to me.

cloud-fan commented on core/src/main/scala/org/apache/spark/storage/BlockManagerManagedBuffer.scala:
```diff
@@ -17,11 +17,15 @@
 
 /**
- * This [[ManagedBuffer]] wraps a [[ChunkedByteBuffer]] retrieved from the [[BlockManager]]
+ * This [[ManagedBuffer]] wraps a ManagedBuffer retrieved from the [[BlockManager]]
```
`wraps a [[BlockData]]`

cloud-fan commented on core/src/main/scala/org/apache/spark/storage/BlockManagerManagedBuffer.scala:
```diff
@@ -31,17 +35,31 @@ import org.apache.spark.util.io.ChunkedByteBuffer
+    data: BlockData,
+    dispose: Boolean) extends ManagedBuffer {
+
+  private val refCount = new AtomicInteger(1)
```
maybe we should mention it in the class doc that the `BlockData` will be disposed automatically via reference count.

cloud-fan commented on core/src/main/scala/org/apache/spark/storage/BlockManagerManagedBuffer.scala:
```diff
@@ -31,17 +35,31 @@ import org.apache.spark.util.io.ChunkedByteBuffer
     blockId: BlockId,
-    chunkedBuffer: ChunkedByteBuffer) extends NettyManagedBuffer(chunkedBuffer.toNetty) {
+    data: BlockData,
+    dispose: Boolean) extends ManagedBuffer {
```
`needDispose` may be a better name

vanzin replied:
Hmm, I prefer `dispose`, because it's not about needing to dispose the buffer, but wanting to dispose the buffer.

cloud-fan commented on core/src/main/scala/org/apache/spark/storage/BlockManager.scala:
```diff
@@ -1065,7 +1084,7 @@ private[spark] class BlockManager(
             replicate(blockId, bytesToReplicate, level, remoteClassTag)
           } finally {
-            bytesToReplicate.unmap()
+            bytesToReplicate.dispose()
```
why change `unmap` to `dispose`?

vanzin replied:
Because there's no `BlockData.unmap()`.

cloud-fan replied:
I'm afraid this may counteract the effort we made in https://github.com/apache/spark/pull/16499

Ideally `unmap` and `dispose` do different things

cc @mallman

vanzin replied:
`BlockData.dispose` calls `ChunkedByteBuffer.unmap`.

cloud-fan commented on core/src/main/scala/org/apache/spark/storage/DiskStore.scala:
```diff
@@ -17,48 +17,67 @@
-        Closeables.close(fileOutputStream, threwException)
+        out.close()
+      } catch {
+        case ioe: IOException =>
```
why this? `threwException` starts with `true`

vanzin replied:
The code needs to catch any exception thrown by `out.close()` and also remove the block in that case. That wasn't done before.

cloud-fan commented on core/src/main/scala/org/apache/spark/storage/BlockManager.scala:
```diff
@@ -56,6 +57,44 @@ private[spark] class BlockResult(
+
+  def toInputStream(): InputStream
+
+  def toNetty(): Object
```
why the return type is `Object`?

vanzin replied:
See `ManagedBuffer.convertToNetty()`.

Commit:
Feedback.

SparkQA commented:
**[Test build #75120 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/75120/testReport)** for PR 17295 at commit [`00b6d00`](https://github.com/apache/spark/commit/00b6d00b35fef1df4492288919ae54302afae8cb).

SparkQA commented:
**[Test build #75120 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/75120/testReport)** for PR 17295 at commit [`00b6d00`](https://github.com/apache/spark/commit/00b6d00b35fef1df4492288919ae54302afae8cb).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/75120/
Test PASSed.

cloud-fan commented on core/src/main/scala/org/apache/spark/storage/BlockManager.scala:
```diff
@@ -56,6 +57,49 @@ private[spark] class BlockResult(
+
+  override def size: Long = buffer.size
+
+  override def dispose(): Unit = buffer.unmap()
```
can we define the semantic of the `BlockData.dispose` clearly? It's quite confusing here that the `dispose` method call `buffer.unmap` while `ChunkedByteBuffer` also has a `dispose` method.

vanzin replied:
I think `BlockData.dispose()` is pretty well defined. "Release any resources held by the object." What's confusing is that there's both `dispose()` and `unmap()` in `ChunkedByteBuffer`, when there used to be only `dispose()`. It's confusing to have two different methods for releasing resources, and that confusion is not being caused by this patch.

`BlockData` is not just a wrapper around `ChunkedByteBuffer`; if it were there wouldn't be a need for it. Which is why calling the method `unmmap()` wouldn't make any sense here, since that's very specific to memory-mapped byte buffers.

vanzin replied:
BTW I'm really starting to think the fix in #16499, while technically correct, is more confusing that it should be. The problem is not that the code was disposing of off-heap buffers; the problem is that buffers read from the memory store should not be disposed of, while buffers read from the disk store should.

So it's not really a matter of dispose vs. unmap, but a matter of where the buffer come from. (Which is kinda what I had in this patch with the `autoDispose` parameter to `ByteBufferBlockData`. Perhaps I should revive that and get rid of `StorageUtils.unmap`, which is just confusing.)

Commit:
Remove "StorageUtils.unmap()".

It is not the type of buffer that defines whether it should be disposed
or not, but rather where it comes from: if it comes from the memory store,
it should not be disposed. Any other case (disk store, temporary serialized
buffers, etc), the buffer should be disposed.

It just happened that this sort of aligned with the types (buffers from the
memory store are normal buffers, buffers from the disk store are mapped buffers
in certain cases). But the origin defines who owns the buffer and, thus, who
should dispose of it.

vanzin commented:
I removed `StorageUtils.unmap()` in my last commit (see commit message for details). That makes the confusion go away.

The replication tests fail from time to time but they seem to be flaky without this patch. See;
https://spark-tests.appspot.com/test-details?suite_name=org.apache.spark.storage.BlockManagerProactiveReplicationSuite&test_name=proactive+block+replication+-+5+replicas+-+4+block+manager+deletions

SparkQA commented:
**[Test build #75226 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/75226/testReport)** for PR 17295 at commit [`d4013f9`](https://github.com/apache/spark/commit/d4013f9453ca4d7c4a2462c420826da46f514fcc).

SparkQA commented:
**[Test build #75226 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/75226/testReport)** for PR 17295 at commit [`d4013f9`](https://github.com/apache/spark/commit/d4013f9453ca4d7c4a2462c420826da46f514fcc).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/75226/
Test PASSed.

cloud-fan commented on core/src/main/scala/org/apache/spark/storage/BlockManagerManagedBuffer.scala:
```diff
@@ -17,31 +17,52 @@
  * so that the corresponding block's read lock can be released once this buffer's references
  * are released.
  *
+ * If `dispose` is set to try, the [[BlockData]]will be disposed when the buffer's reference
```
`is set to try` -> `is set to true`

cloud-fan commented:
LGTM, cc @mallman to check the `unmap` part

Commit:
Fix typo.

SparkQA commented:
**[Test build #75267 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/75267/testReport)** for PR 17295 at commit [`ab4b5dd`](https://github.com/apache/spark/commit/ab4b5dd7191c9f73f5419180cedd3051621bd83a).

SparkQA commented:
**[Test build #75267 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/75267/testReport)** for PR 17295 at commit [`ab4b5dd`](https://github.com/apache/spark/commit/ab4b5dd7191c9f73f5419180cedd3051621bd83a).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/75267/
Test PASSed.

Commit:
Merge branch 'master' into SPARK-19556

SparkQA commented:
**[Test build #75323 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/75323/testReport)** for PR 17295 at commit [`4a39cb2`](https://github.com/apache/spark/commit/4a39cb23dae86d4289bf529c27dae21680620cab).

SparkQA commented:
**[Test build #75323 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/75323/testReport)** for PR 17295 at commit [`4a39cb2`](https://github.com/apache/spark/commit/4a39cb23dae86d4289bf529c27dae21680620cab).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/75323/
Test PASSed.

cloud-fan commented:
thanks, merging to master!

mallman commented:
> LGTM, cc @mallman to check the unmap part

LGTM, too. Sorry for the late reply... I've been away the past two weeks.
